scipy/stats/_qmc.py

```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
         return samples  # type: ignore[return-value]
 
 
+class OptimalLatinHypercube(QMCEngine):
+    """Optimal Latin hypercube sampling (OLHS).
+
+    Optimize the sample by doing random permutations of coordinates to lower
+    the centered discrepancy [1]_. The resulting sample is still a LHS.
+
+    Parameters
+    ----------
+    d : int
+        Dimension of the parameter space.
+    start_sample : array_like (n, d), optional
+        Initial sample to optimize. `LatinHypercube`
+        is used to generate a first design otherwise.
+    n_perturbations : int, optional
+        Number of perturbation to perform. Default is 10000.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used.
+
+    Notes
+    -----
+    `n_perturbations` set of random permutations are performed aginst a LHS
+    sample. The best design based on the centered discrepancy is constantly
+    update. Centered discrepancy-based design shows better space filling
+    robustness toward 2D and 3D subprojections. Distance-based design shows
+    better space filling but less robustness to subprojections [2]_.
+
+    References
+    ----------
+    .. [1] Jin et al. "An efficient algorithm for constructing optimal design
+       of computer experiments", Journal of Statistical Planning and
+       Inference, 2005.
+    .. [2] Damblin et al., "Numerical studies of space filling designs:
+       optimization of Latin Hypercube Samples and subprojection properties."
+       Journal of Simulation, 2013.
+
+    Examples
+    --------
+    Generate samples from an optimal design.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalLatinHypercube(d=2)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.84052691, 0.53664833],  # random
+           [0.1545328 , 0.36265316],
+           [0.52177809, 0.93343721],
+           [0.68033825, 0.06474907],
+           [0.26544879, 0.61163943]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.01602810655510689  # random
+
+    You can possibly improve the quality of the sample by performing more
+    optimization iterations by using `niter`:
+
+    >>> sampler_2 = qmc.OptimalLatinHypercube(d=2, n_perturbations=5000)
+    >>> sample_2 = sampler_2.random(n=5)
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648  # random
+
+    """
+
+    def __init__(self, d: IntNumber, *, start_sample: npt.ArrayLike = None,
+                 n_perturbations: IntNumber = 10000,
+                 seed: SeedType = None):
```

Comment:
Is there supposed to be a space around the equals sign when used with types? 
(Do I understand that type information can be moved to a separate file to avoid cluttering the code?)

Reply:
I know it looks strange but yes there is.

Also yes, the types could be moved to a stub file (same name with a different extension `pyi`). It was actually the case in the first version, and then we said that it would be better to not duplicate all the files like this and have the types inline.